File: vision_fullcam/tracking/simple_tracker.py
# vision/tracking/simple_tracker.py
from dataclasses import dataclass
from typing import Dict, List, Tuple
import math

@dataclass
class Tracked:
    track_id: int
    label: str
    bbox: Tuple[int,int,int,int]
    score: float

def iou(a, b) -> float:
    ax1, ay1, ax2, ay2 = a
    bx1, by1, bx2, by2 = b
    ix1, iy1 = max(ax1, bx1), max(ay1, by1)
    ix2, iy2 = min(ax2, bx2), min(ay2, by2)
    iw, ih = max(0, ix2-ix1), max(0, iy2-iy1)
    inter = iw*ih
    area_a = (ax2-ax1)*(ay2-ay1)
    area_b = (bx2-bx1)*(by2-by1)
    union = max(1, area_a + area_b - inter)
    return inter/union

class SimpleTracker:
    """
    IoU 기반. MVP에서는 충분.
    사람이 많아지면 ByteTrack으로 교체(인터페이스 유지).
    """
    def __init__(self, iou_thr: float = 0.3):
        self.iou_thr = iou_thr
        self.next_id = 1
        self.prev: Dict[int, Tracked] = {}

    def update(self, detections: List[Tracked]) -> Dict[int, Tracked]:
        # label별로 매칭
        new_prev: Dict[int, Tracked] = {}
        used = set()

        for tid, prev_t in self.prev.items():
            best_j, best_iou = None, 0.0
            for j, d in enumerate(detections):
                if j in used: 
                    continue
                if d.label != prev_t.label:
                    continue
                s = iou(prev_t.bbox, d.bbox)
                if s > best_iou:
                    best_iou = s
                    best_j = j
            if best_j is not None and best_iou >= self.iou_thr:
                used.add(best_j)
                d = detections[best_j]
                new_prev[tid] = Tracked(tid, d.label, d.bbox, d.score)

        # 신규 트랙 부여
        for j, d in enumerate(detections):
            if j in used: 
                continue
            tid = self.next_id
            self.next_id += 1
            new_prev[tid] = Tracked(tid, d.label, d.bbox, d.score)

        self.prev = new_prev
        return new_prev

File: frontend/mobile/src/vision/tracking/simpleTracker.ts
import type { BBox, Tracked } from '../types';

function iou(a: BBox, b: BBox): number {
  const ix1 = Math.max(a[0], b[0]);
  const iy1 = Math.max(a[1], b[1]);
  const ix2 = Math.min(a[2], b[2]);
  const iy2 = Math.min(a[3], b[3]);
  const iw = Math.max(0, ix2 - ix1);
  const ih = Math.max(0, iy2 - iy1);
  const inter = iw * ih;
  const areaA = (a[2] - a[0]) * (a[3] - a[1]);
  const areaB = (b[2] - b[0]) * (b[3] - b[1]);
  const union = Math.max(1, areaA + areaB - inter);
  return inter / union;
}

/**
 * IoU 기반 심플 트래커.
 * Python simple_tracker.py 1:1 포팅.
 */
export class SimpleTracker {
  private iouThr: number;
  private nextId: number;
  private prev: Map<number, Tracked>;

  constructor(iouThr: number = 0.3) {
    this.iouThr = iouThr;
    this.nextId = 1;
    this.prev = new Map();
  }

  update(detections: Tracked[]): Map<number, Tracked> {
    const newPrev = new Map<number, Tracked>();
    const used = new Set<number>();

    // 기존 트랙에 매칭
    for (const [tid, prevT] of this.prev) {
      let bestJ: number | null = null;
      let bestIoU = 0.0;

      for (let j = 0; j < detections.length; j++) {
        if (used.has(j)) continue;
        const d = detections[j];
        if (d.label !== prevT.label) continue;

        const s = iou(prevT.bbox, d.bbox);
        if (s > bestIoU) {
          bestIoU = s;
          bestJ = j;
        }
      }

      if (bestJ !== null && bestIoU >= this.iouThr) {
        used.add(bestJ);
        const d = detections[bestJ];
        newPrev.set(tid, {
          trackId: tid,
          label: d.label,
          bbox: d.bbox,
          score: d.score,
        });
      }
    }

    // 신규 트랙 부여
    for (let j = 0; j < detections.length; j++) {
      if (used.has(j)) continue;
      const d = detections[j];
      const tid = this.nextId++;
      newPrev.set(tid, {
        trackId: tid,
        label: d.label,
        bbox: d.bbox,
        score: d.score,
      });
    }

    this.prev = newPrev;
    return newPrev;
  }

  reset(): void {
    this.prev = new Map();
    this.nextId = 1;
  }
}
